File: database-driver-plugin/src/main/java/de/lino/database/database/sql/SQLDatabaseProvider.java
package de.lino.database.database.sql;

import de.lino.database.database.AbstractCachedDatabaseSection;
import de.lino.database.database.AbstractLazyDatabaseProvider;
import de.lino.database.database.DatabaseProvider;
import de.lino.database.database.DatabaseSection;
import de.lino.database.database.DatabaseType;
import de.lino.database.database.SectionConfig;
import org.jetbrains.annotations.NotNull;

import java.sql.SQLException;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;

/**
 * The shared {@link DatabaseProvider} implementation behind every SQL vendor this driver
 * supports (MySQL, PostgreSQL, MariaDB, SQLite, H2, Oracle, Microsoft SQL Server, Apache Derby -
 * see the vendor-specific subclasses in the sibling packages), each {@link DatabaseSection}
 * mapping to one table, all sharing this database's single {@link SQLExecution} connection pool.
 * Section lifecycle and caching live in {@link AbstractLazyDatabaseProvider}; this class only
 * supplies the vendor-aware storage operations - listing tables, constructing a
 * {@link SQLDatabaseSection}, dropping a table.
 */
public class SQLDatabaseProvider extends AbstractLazyDatabaseProvider {

    /**
     * The SQL vendor this database is connected to, needed to pick the right table-listing
     * query in {@link #getPattern} and the right BLOB column type per {@link SQLDatabaseSection}.
     */
    private final DatabaseType databaseType;

    /**
     * The connection pool shared by this database and every {@link SQLDatabaseSection} it creates.
     */
    private final SQLExecution sqlExecution;

    /**
     * Connects via {@code sqlExecution} and discovers every existing table's name of
     * {@code databaseType}. Only names - no section objects, no row data - so construction
     * cost is one table-listing query, independent of how much the database holds.
     *
     * @param databaseType  the SQL vendor being connected to
     * @param sqlExecution  the connection pool to run every query and update through
     */
    public SQLDatabaseProvider(@NotNull DatabaseType databaseType, @NotNull SQLExecution sqlExecution) {

        this.databaseType = databaseType;
        this.sqlExecution = sqlExecution;

        this.reload();

    }

    /**
     * The connection pool this database runs every query and update through - exposed (since
     * 1.3.16) as the deliberate raw-SQL escape hatch for a consumer feature that genuinely
     * cannot be expressed over {@link de.lino.database.database.DatabaseSection}'s
     * key/value surface (the motivating case: a Postgres {@code tsvector}/GIN full-text search
     * index, which needs vendor-specific column types and index DDL no generic section can
     * carry). Callers share this pool with every section of this database: never call
     * {@link SQLExecution#shutdown()} on it - the provider owns its lifecycle - and keep
     * statements short-lived so section traffic is never starved of pooled connections.
     *
     * @return the shared connection pool
     */
    @NotNull
    public SQLExecution getSqlExecution() {
        return this.sqlExecution;
    }

    @Override
    public void shutdown() {
        this.sqlExecution.shutdown();
        this.forgetSections();
    }

    /**
     * {@inheritDoc}
     * <p>
     * Runs {@link #getPattern}'s vendor-specific table-listing query, streaming each
     * {@code TABLE_NAME} to {@code consumer}.
     */
    @Override
    protected void discoverNames(@NotNull Consumer<String> consumer) {

        this.sqlExecution.executeQuery(getPattern(this.databaseType), resultSet -> {

            try {

                while (resultSet.next()) consumer.accept(resultSet.getString("TABLE_NAME"));

            } catch (final SQLException exception) {
                exception.printStackTrace();
            }

            return true;
        }, true);

    }

    @Override
    protected AbstractCachedDatabaseSection constructSection(@NotNull String name, @NotNull SectionConfig config) {
        return new SQLDatabaseSection(this.databaseType, name, this.sqlExecution, config);
    }

    @Override
    protected void dropSectionRemote(@NotNull String name) {
        this.sqlExecution.executeUpdate("DROP TABLE " + name);
    }

    /**
     * {@link DatabaseSection#clear() Clears} every section concurrently rather than one table at
     * a time, since every section shares the same {@link SQLExecution} connection pool - which is
     * itself built for concurrent multi-threaded use - so clearing them in parallel is no less
     * safe than clearing them sequentially, just faster; overrides {@link DatabaseProvider}'s
     * default, which would otherwise run the whole sequential {@link #clear()} on a single
     * background thread.
     *
     * @return a {@link CompletableFuture} that completes once every section has been cleared
     */
    @Override
    public CompletableFuture<Void> clearAsync() {

        final List<CompletableFuture<Void>> pending = this.getSections().stream().map(DatabaseSection::clearAsync).toList();

        return CompletableFuture.allOf(pending.toArray(CompletableFuture[]::new)).thenRun(this::forgetSections);

    }

    /**
     * Builds the vendor-specific query that lists every existing table's name as
     * {@code TABLE_NAME}, used by {@link #discoverNames} to enumerate this database's sections.
     *
     * @param databaseType the SQL vendor to build a table-listing query for
     * @return the vendor-specific table-listing query
     */
    private static @NotNull String getPattern(@NotNull DatabaseType databaseType) {

        String tablePattern;

        switch (databaseType) {

            case SQLITE -> tablePattern = "SELECT name AS TABLE_NAME FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%';";
            case APACHE_DERBY -> tablePattern = "SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='App' AND TABLE_NAME NOT LIKE 'sqlite_%';";
            case MICROSOFT_SQL_SERVER -> tablePattern = "SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='dbo'";
            case ORACLE -> tablePattern = "SELECT table_name AS TABLE_NAME FROM all_tables WHERE owner='SCHEMA_NAME'";
            default -> tablePattern = "SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='PUBLIC'";

        }
        return tablePattern;
    }

}

File: python/database-driver-plugin/src/database_driver/plugin/database/sql/sql_database_provider.py
"""Mirror of ``de.lino.database.database.sql.SQLDatabaseProvider``."""

from __future__ import annotations

import asyncio
from collections.abc import Callable
from typing import Any

from database_driver.api.database.database_type import DatabaseType
from database_driver.api.database.section_config import SectionConfig

from database_driver.plugin.database.abstract_cached_database_section import AbstractCachedDatabaseSection
from database_driver.plugin.database.abstract_lazy_database_provider import AbstractLazyDatabaseProvider
from database_driver.plugin.database.sql.sql_database_section import SQLDatabaseSection
from database_driver.plugin.database.sql.sql_execution import SQLExecution


class SQLDatabaseProvider(AbstractLazyDatabaseProvider):
    """The shared ``DatabaseProvider`` implementation behind every SQL vendor this
    driver supports (MySQL, PostgreSQL, MariaDB, SQLite, Oracle, Microsoft SQL Server -
    see the vendor-specific subclasses in the sibling packages), each section mapping to
    one table, all sharing this database's single ``SQLExecution`` connection pool.
    Section lifecycle and caching live in ``AbstractLazyDatabaseProvider``; this class
    only supplies the vendor-aware storage operations - listing tables, constructing a
    ``SQLDatabaseSection``, dropping a table."""

    def __init__(self, database_type: DatabaseType, sql_execution: SQLExecution) -> None:
        super().__init__()
        self.database_type = database_type
        self._sql_execution = sql_execution
        self.reload()

    def get_sql_execution(self) -> SQLExecution:
        """The connection pool this database runs every query and update through -
        exposed as the deliberate raw-SQL escape hatch for a consumer feature that
        genuinely cannot be expressed over ``DatabaseSection``'s key/value surface (the
        motivating case: a Postgres ``tsvector``/GIN full-text search index, which needs
        vendor-specific column types and index DDL no generic section can carry).
        Callers share this pool with every section of this database: never call
        ``SQLExecution.shutdown()`` on it - the provider owns its lifecycle - and keep
        statements short-lived so section traffic is never starved of pooled
        connections."""
        return self._sql_execution

    def shutdown(self) -> None:
        self._sql_execution.shutdown()
        self.forget_sections()

    def discover_names(self, consumer: Callable[[str], None]) -> None:
        """Runs the vendor-specific table-listing query, streaming each table name to
        ``consumer``."""

        def collect(cursor: Any) -> bool:
            for row in cursor.fetchall():
                consumer(row[0])
            return True

        self._sql_execution.execute_query(_pattern(self.database_type), collect, True)

    def construct_section(self, name: str, config: SectionConfig) -> AbstractCachedDatabaseSection:
        return SQLDatabaseSection(self.database_type, name, self._sql_execution, config)

    def drop_section_remote(self, name: str) -> None:
        self._sql_execution.execute_update(f"DROP TABLE {name}")

    async def clear_async(self) -> None:
        """Clears every section concurrently rather than one table at a time, since
        every section shares the same ``SQLExecution`` connection pool - which is itself
        built for concurrent multi-threaded use - so clearing them in parallel is no less
        safe than clearing them sequentially, just faster; overrides
        ``DatabaseProvider``'s default, which would otherwise run the whole sequential
        :meth:`clear` on a single worker thread."""
        await asyncio.gather(*(section.clear_async() for section in self.get_sections()))
        self.forget_sections()


def _pattern(database_type: DatabaseType) -> str:
    """Builds the vendor-specific query that lists every existing table's name, used by
    ``discover_names`` to enumerate this database's sections.

    Two deliberate deviations from the Java edition's patterns: PostgreSQL's
    ``information_schema`` stores schema names lowercase (``'public'``, not
    ``'PUBLIC'``), and MySQL/MariaDB scope tables by database (``DATABASE()``), not by a
    ``PUBLIC`` schema - the Java queries return empty on those vendors.
    """
    if database_type is DatabaseType.SQLITE:
        return "SELECT name FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%'"
    if database_type is DatabaseType.MICROSOFT_SQL_SERVER:
        return "SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='dbo'"
    if database_type is DatabaseType.ORACLE:
        return "SELECT table_name FROM user_tables"
    if database_type is DatabaseType.POSTGRES_SQL:
        return "SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='public'"
    if database_type in (DatabaseType.MY_SQL, DatabaseType.MARIA_DB):
        return "SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA=DATABASE()"
    return "SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='PUBLIC'"
